### PR TITLE
Feature mrmr

### PR DESCRIPTION
**Pull request by fabioscantamburlo**
# Description

Implement MRMR feature selection model following the article:
[MrMr By S. Mazzanti](https://towardsdatascience.com/mrmr-explained-exactly-how-you-wished-someone-explained-to-you-9cf4ed27458b
)
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Remaining:
- [x] User guide.
- [x] Minor fixes. 

# Checklist:

- [x] My code follows the style guidelines (flake8)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (also to the readme.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added tests to check whether the new feature adheres to the sklearn convention
- [x] New and existing unit tests pass locally with my changes

If you feel your PR is ready for a review, ping @koaning or @mbrouns, @FBruzzesi

### CODE REVIEW

File: sklego/feature_selection/mrmr.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    Xl = X_norm[:, left]
+
+    num = (Xl[:, None, :] * Xs[:, :, None]).sum(axis=0)
+    den = np.sqrt((Xl[:, None, :] ** 2).sum(axis=0)) * np.sqrt((Xs[:, :, None] ** 2).sum(axis=0))
+
+    return np.sum(np.abs(np.nan_to_num(num / den, nan=np.finfo(float).eps)), axis=0)
+
+
+class MaximumRelevanceMinimumRedundancy(SelectorMixin, BaseEstimator):
+    """Maximum Relevance Minimum Redundancy (MRMR) is an iterative feature selection method commonly used in data
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:42]: I would add the formula in the docstring as well, something along the line:

$\text{score}_{i}(f) = \frac{\text{relevance}(f | y}{\text{reduncancy}(f | \text{selected}_{i-1}}$
for each feature $f$ is

Edit: Not sure why github doesn't like the formula :|

File: sklego/feature_selection/mrmr.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    number of features is reached.
+
+    Parameters
+    ----------
+    k : int
+        Number of feature the model should use.
+    relevance_func : str | Callable,, default= "f"(f_classif or  f_regression from sklearn.feature_selection)
+        [f_classif](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html)
+        [f_regression](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html)
+        The relevance function to use.
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:69]: How about:

```suggestion
    relevance_func : str | Callable, default="f"
       The relevance function to use. The default maps to scikit-learn [f_classif](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html) or  [f_regression](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html) for classification or regression (resp.)
```
?

File: sklego/feature_selection/mrmr.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    redundancy_func : str | Callable, default="p" (Pearson correlation)
+        The redundancy function to use.
+    kind : Literal["auto", "classficiation", "regression"], default="auto".
+        'classification' or 'regression' or 'auto' if auto the model
+        will try to infer the type of problem looking at the y data type, by default "auto".
+
+    !! warning:
+        If a custom relevance_func is provided it must have this signature:
+        Callable[[np.ndarray, np.ndarray], np.ndarray]
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:78]:
```suggestion
        `Callable[[np.ndarray, np.ndarray], np.ndarray]`
```

File: sklego/feature_selection/mrmr.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        It should accept X, y as arguments and it should compute the score for each feature of X
+        and return an array of shape (n_features_in_,).
+    !! warning:
+        If a custom redundancy_func is provided it must have the same signature as the method _redundancy_pearson
+
+    Attributes
+    ----------
+    _y_dtype : data type of y
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:86]:
```suggestion
    _y_dtype : np.dtype
        data type of y
```

File: sklego/feature_selection/mrmr.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        elif callable(self.redundancy_func):
+            return self.redundancy_func
+        else:
+            raise ValueError(f"Redundancy function supported are 'p' or Callable, got {self.redundancy_func}")
+
+    def fit(self, X, y):
+        """Fit the underlying feature selection algorithm on the training data `X` and `y`
+        using the provided redundancy and relevance functions.
+
+
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:164]:
```suggestion

```

File: sklego/feature_selection/mrmr.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        -------
+        self : MaximumRelevanceMinimumRedundancy
+            The fitted estimator.
+
+        Raises
+        ------
+        ValueError
+            if:
+
+                k parameter is not integer type or is < n_features_in (X.shape[1]) or < 1
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:182]: Formatting

File: sklego/feature_selection/mrmr.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        List of indexes of the left features at iteration i-th. Mrmr will select a feature
+        from this list.
+
+    Returns
+    -------
+    np.ndarray, shape = (len(left), )
+        The array containing the redundancy score using pearson correlation.
+    """
+    # if len(selected) == 0:
+    #     return np.ones(len(left))
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:29]: I personally would keep this edge case even though in the code below can never happen

File: docs/_scripts/feature-selection.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
+from pathlib import Path
+
+_file = Path(__file__)
+print(f"Executing {_file}")
+
+_static_path = Path("_static") / _file.stem
+_static_path.mkdir(parents=True, exist_ok=True)
+
+import matplotlib.pyplot as plt
```
Comment (FBruzzesi) [inline docs/_scripts/feature-selection.py:9]: I know this is WIP, but can we use seaborn style as in the other plots?

```suggestion
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme()
```

File: sklego/feature_selection/mrmr.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+    Examples
+    --------
+    ```py
+    from sklego.feature_selection.mrmr import MaximumRelevanceMinimumRedundancy
+
+    mrmr =  MaximumRelevanceMinimumRedundancy(k=4,
+            kind='auto',
+            redundancy_func='p',
+            relevance_func='f')
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:100]: Formatting

File: sklego/feature_selection/mrmr.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    _y_dtype : data type of y
+    selected_features_ : array-like of shape (k,)
+        Indexes of the selected features.
+    scores_ : array-like of shape (k,)
+        Scores of the selected features.
+
+    Examples
+    --------
+    ```py
+    from sklego.feature_selection.mrmr import MaximumRelevanceMinimumRedundancy
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:95]: I would prefer to have a similar logic as in the other submodules. Namely:

- In `sklego.feature_selection.__init__.py` add the import for `MaximumRelevanceMinimumRedundancy`
- Then one is able to import from feature selection module
     ```
     from sklego.feature_selection import MaximumRelevanceMinimumRedundancy
     ```

File: sklego/feature_selection/mrmr.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+    mrmr =  MaximumRelevanceMinimumRedundancy(k=4,
+            kind='auto',
+            redundancy_func='p',
+            relevance_func='f')
+
+    X, y = ...
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:102]: Possibly add an example so that people can copy paste a working implementation.
Using any of the non-raising test should be enough

**Review by FBruzzesi — COMMENTED** [review]
I left a few comments on style, formatting, phrasing and so on..

@koaning as I was somewhat involved in the code and logic implementation, I may be biased in reviewing it 😁

File: sklego/feature_selection/mrmr.py (hunk 1 above)
Comment (koaning) [reply sklego/feature_selection/mrmr.py:42]: Maybe this? 

$` \text{score}_{i}(f) = \frac{\text{relevance}(f | y)}{\text{reduncancy}(f | \text{selected}_{i-1})} `$

File: docs/_scripts/feature-selection.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+X, y = X_pd.to_numpy(), y_pd.to_numpy()
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=10000, random_state=42)
+X_train = X_train.reshape(60000, 28 * 28)
+X_test = X_test.reshape(10000, 28 * 28)
+
+
+from scipy.spatial.distance import cosine
+
+
+def _redundancy_cosine_scipy(X, selected, left):
```
Comment (koaning) [inline docs/_scripts/feature-selection.py:36]: Since it's public facing, maybe no underscore here?

```suggestion
def redundancy_cosine_scipy(X, selected, left):
```

Also would need to change below.

File: tests/test_feature_selection/test_mrmr.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    redundancy_scores = _redundancy_pearson(X, selected, left)
+    redundancy_scores_expected = np.array([2 * 1.0, 2 * 1.0])
+    assert np.allclose(redundancy_scores, redundancy_scores_expected, atol=0.00001)
+
+
+@pytest.mark.parametrize("k", [1.2, 5, -1])
+def test_mrmr_raises(k):
+    X, y = make_classification(n_features=4)
+    with pytest.raises(ValueError):
+        _ = MaximumRelevanceMinimumRedundancy(k=k).fit(X, y)
```
Comment (koaning) [inline tests/test_feature_selection/test_mrmr.py:40]: Nitpick, but why have the `_` here? 

```suggestion
        MaximumRelevanceMinimumRedundancy(k=k).fit(X, y)
```

File: docs/_scripts/feature-selection.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
+from pathlib import Path
```
Comment (koaning) [inline docs/_scripts/feature-selection.py:1]: Isn't there a markdown file missing for the docs that's supposed to use this script?

**Review by koaning — COMMENTED** [review]
Left some quick notes, the main thing seems to be a missing documentation file.

File: docs/_scripts/feature-selection.py (hunk 12 above)
Comment (fabioscantamburlo) [reply docs/_scripts/feature-selection.py:36]: I'm removing this anyway, I'll clean a bit the script removing unnecessary code.

File: sklego/feature_selection/mrmr.py (hunk 7 above)
Comment (fabioscantamburlo) [reply sklego/feature_selection/mrmr.py:29]: This will prevent the code the be tested 100% because it's never hit.
What do you think?

Comment (FBruzzesi) [reply sklego/feature_selection/mrmr.py:29]: We are not aiming to achieve 100% coverage necessarily.
If you want to keep the edge cases _and_ the coverage, you may add a unit test specific for this function

File: sklego/feature_selection/mrmr.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+        'classification' or 'regression' or 'auto' if auto the model
+        will try to infer the type of problem looking at the y data type, by default "auto".
+
+    !! warning:
+        If a custom relevance_func is provided it must have this signature:
+        `Callable[[np.ndarray, np.ndarray], np.ndarray]`
+        It should accept X, y as arguments and it should compute the score for each feature of X
+        and return an array of shape (n_features_in_,).
+    !! warning:
+        If a custom redundancy_func is provided it must have the same signature as the method _redundancy_pearson
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:82]: Couple of considerations here:
- Both admonitions are not rendering properly (see suggestion below)
- I would consider moving them above, between the formula and the parameters, instead of between parameters and attributes
- The second warning may not be clear to a user as `_redundancy_pearson` cannot be checked if not in the code directly, maybe describe such signature (`X, selected, left`) directly

```suggestion
    !!! warning
        If a custom relevance_func is provided it must have this signature:
        `Callable[[np.ndarray, np.ndarray], np.ndarray]`
        It should accept X, y as arguments and it should compute the score for each feature of X
        and return an array of shape (n_features_in_,).
    
    !!! warning
        If a custom redundancy_func is provided it must have the same signature as the method _redundancy_pearson
```

File: sklego/feature_selection/mrmr.py [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+        data type of y
+    selected_features_ : array-like of shape (k,)
+        Indexes of the selected features.
+    scores_ : array-like of shape (k,)
+        Scores of the selected features.
+
+    Examples
+    --------
+    ```py
+    from sklego.feature_selection.mrmr import MaximumRelevanceMinimumRedundancy
```
Comment (FBruzzesi) [inline sklego/feature_selection/mrmr.py:96]: Can import directly from `feature_selection` module:

```suggestion
    from sklego.feature_selection import MaximumRelevanceMinimumRedundancy
```

File: .gitignore [hunk 17]
```diff
@@ -120,4 +121,7 @@ dmypy.json
 .DS_Store
 
 # Local Netlify folder
-.netlify
\ No newline at end of file
+.netlify
+
+# venv
+venv*
```
Comment (FBruzzesi) [inline .gitignore:127]: Isn't this just duplicated?

**Review by FBruzzesi — COMMENTED** [review]
A lot of tiny details

**Comment by fabioscantamburlo** [comment]
Gonna fix them asap 🙏

File: docs/user-guide/feature-selection.md [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+F1 score: 0.854
+Feature selection method: mutual_info
+F1 score: 0.879
+Feature selection method: mrmr
+F1 score: 0.925
+Feature selection method: mrmr_smile
+F1 score: 0.849
+```
+
+As expected MRMR feature selection model provides better results compared against the other methods. (Smile still performs very well!)
```
Comment (koaning) [inline docs/user-guide/feature-selection.md:60]: Nitpick, but maybe this is friendlier:. 

```suggestion
The MRMR feature selection model provides better results compared against the other methods, although the smile technique performs rather good as well. 
```

**Review by koaning — COMMENTED** [review]
I just have one nitpick. @FBruzzesi unless you have a strong opinion I'd say we merge this one in.

**Review by koaning — APPROVED** [review]
I'll merge it for now. Release will have to wait for the hierarchical stuff, but yeah ... this'll be a nice addition for the v0.8.0 release 😄

**Comment by fabioscantamburlo** [comment]
Great! Thanks for reviewing it 🙂

**Comment by FBruzzesi** [comment]
> this'll be a nice addition for the v0.8.0 release 😄

@koaning do you think we should consider start adding tags for when a new feature is introduced? I would vote for it as we are rolling out features with a faster pace than the last few years 😁
